internal/provider/provider.go

```diff
@@ -27,12 +27,16 @@ import (
 type provider struct {
 	logger hclog.Logger
 	cache  map[cacheKey]*api.Secret
+
+	// Allows mocking Kubernetes API for tests.
+	k8sClientConfig func() (*rest.Config, error)
```

Comment:
Have you considered using the fake clientset that's part of client-go? I think it might simplify some of the mock server setup that's required here. To use it you can set a k8s client var to a type of kubernetes.Interface, and use a fake clientset for tests and a real clientset otherwise. There's an example in the vault-k8s tests now: https://github.com/hashicorp/vault-k8s/blob/0c69acbf33605f66cd1316fd9cd5380e7560f84d/helper/cert/source_gen_test.go#L160

Reply:
Oh nice! TIL, thanks